internal/client/client.go

```diff
@@ -0,0 +1,232 @@
+
+// apiError converts an API error 400 JSON to a status response
+//
+// The response body should alway be of the form:
+//     {"status": false, "errors": [{"id": ERROR_ID, "msg": ERROR_MESSAGE}, ...]}
+func apiError(resp *http.Response) *APIResponse {
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return clientError(err)
+	}
+	return NewAPIResponse(body)
+}
+
+// GetRaw returns raw data from a GET request
+// Errors from the client and from the API are returned as an APIResponse
+func GetRaw(socket, method, path string) ([]byte, *APIResponse) {
+	resp, err := Request(socket, method, path, "", map[string]string{})
+	if err != nil {
+		return nil, clientError(err)
+	}
+
+	// Convert the API's JSON error response to an error type and return it
+	if resp.StatusCode == 400 {
+		return nil, apiError(resp)
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, clientError(err)
+	}
+
+	return body, nil
+}
+
+// GetJSONAll returns all JSON results from a GET request using offset/limit
+// This function makes 2 requests, the first with limit=0 to get the total number of results,
+// and then with limit=TOTAL to fetch all of the results.
+// The path passed to GetJSONAll should not include the limit or offset query parameters
+// Errors from the client and from the API are returned as an APIResponse
+func GetJSONAll(socket, path string) ([]byte, *APIResponse) {
+	body, err := GetRaw(socket, "GET", path+"?limit=0")
+	if err != nil {
+		return nil, err
+	}
+
+	// We just want the total
+	var j interface{}
+	jerr := json.Unmarshal(body, &j)
+	if jerr != nil {
+		return nil, clientError(jerr)
+	}
+	m := j.(map[string]interface{})
+	var v interface{}
+	var ok bool
+	if v, ok = m["total"]; !ok {
+		return nil, clientError(errors.New("Response is missing the total value"))
+	}
+
+	switch total := v.(type) {
+	case float64:
+		allResults := fmt.Sprintf("%s?limit=%v", path, total)
+		return GetRaw(socket, "GET", allResults)
+	}
+	return nil, clientError(errors.New("Response 'total' is wrong type"))
+}
+
+// PostRaw sends a POST with raw data and returns the raw response body
+// Errors from the client and from the API are returned as an APIResponse
+func PostRaw(socket, path, body string, headers map[string]string) ([]byte, *APIResponse) {
+	resp, err := Request(socket, "POST", path, body, headers)
+	if err != nil {
+		return nil, clientError(err)
+	}
+
+	// Convert the API's JSON error response to an error type and return it
+	if resp.StatusCode == 400 {
+		return nil, apiError(resp)
+	}
+	defer resp.Body.Close()
+
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, clientError(err)
+	}
+
+	return responseBody, nil
+}
+
+// PostTOML sends a POST with TOML data and the Content-Type header set to "text/x-toml"
+// It returns the raw response data or errors as an APIResponse
+func PostTOML(socket, path, body string) ([]byte, *APIResponse) {
```

Comment:
Would be nice to pass the body as `interface{}` here and do the marshalling to TOML in this function (same for JSON below).